libsolidity/analysis/ControlFlowAnalyzer.cpp

```diff
@@ -109,12 +152,13 @@ void ControlFlowAnalyzer::checkUninitializedAccess(CFGNode const* _entry, CFGNod
 		nodeInfo.unassignedVariablesAtExit = std::move(unassignedVariables);
 
 		// Propagate changes to all exits and queue them for traversal, if needed.
-		for (auto const& exit: currentNode->exits)
-			if (
-				auto exists = valueOrNullptr(nodeInfos, exit);
-				nodeInfos[exit].propagateFrom(nodeInfo) || !exists
-			)
-				nodesToTraverse.insert(exit);
+		if (!reverts)
+			for (auto const& exit: currentNode->exits)
+				if (
+					auto exists = valueOrNullptr(nodeInfos, exit);
+					nodeInfos[exit].propagateFrom(nodeInfo) || !exists
```

Comment:
This condition is weird on so many levels. It's either buggy or written in such a way that it's hard to tell if the unusual behavior is intentional or a bug:
- `exists` is a pointer to `NodeInfo` but named and used like a boolean. And `auto` makes it harder to figure out at a glance.
- Uses `nodeInfo[exit]` instead of `*exists` which it has just defined.
- Operator `[]` will insert `exit` into the map if it's not there yet. Is that intentional?
- The `!exists` part looks like it's meant to prevent accessing `nodeInfos` when the pointer is null but it does not do that when it's at the end.

Shouldn't it be rather written like this?
```suggestion
					NodeInfo const* exitNodeInfo = valueOrNullptr(nodeInfos, exit);
					!exitNodeInfo || exitNodeInfo.propagateFrom(nodeInfo)
```

Reply:
> Uses `nodeInfo[exit]` instead of `*exists` which it has just defined.

Well, because it could be null ;)

> Operator `[]` will insert `exit` into the map if it's not there yet. Is that intentional?

Yes, we want to add all exits that don't yet exist

> The `!exists` part looks like it's meant to prevent accessing `nodeInfos` when the pointer is null but it does not do that when it's at the end.

That is very deliberate. We always want to add the node if it doesn't exist. We also want to add the node if it was changed (that is, if `propagateFrom()` returned `true`) because we need to process it again then.
We also always want to call `propagateFrom()` to propagate any changes from the current iteration, no matter if it already exists or not.

I agree that this part is very confusing and I misunderstood it myself the first time I looked at this code (I am not the author) and even fixed a bug in it a few months ago.

Maybe we can find a better way to have the intended behavior and also understand it.

That said, this PR doesn't really touch this code and I don't think any changes to it should be made in this PR.